File: ema/Src/Access/Impl/Attrib.cpp
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license      --
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
 *|                See the project's LICENSE.md for details.                  --
 *|           Copyright Thomson Reuters 2015. All rights reserved.            --
 *|-----------------------------------------------------------------------------
 */

#include "Attrib.h"
#include "FieldList.h"
#include "ElementList.h"
#include "Map.h"
#include "Vector.h"
#include "Series.h"
#include "FilterList.h"
#include "OmmOpaque.h"
#include "OmmXml.h"
#include "OmmAnsiPage.h"
#include "ReqMsg.h"
#include "RefreshMsg.h"
#include "UpdateMsg.h"
#include "StatusMsg.h"
#include "PostMsg.h"
#include "AckMsg.h"
#include "GenericMsg.h"
#include "MsgDecoder.h"
#include "ExceptionTranslator.h"

using namespace thomsonreuters::ema::access;

extern const EmaString& getDTypeAsString( DataType::DataTypeEnum dType );

Attrib::Attrib() :
 _pDecoder( 0 )
{
}

Attrib::~Attrib()
{
}

DataType::DataTypeEnum Attrib::getDataType() const
{
	return _pDecoder->getAttribData().getDataType();
}

const ComplexType& Attrib::getData() const
{
	return static_cast<const ComplexType&>( _pDecoder->getAttribData() );
}

const ReqMsg& Attrib::getReqMsg() const
{
	const Data& attrib = _pDecoder->getAttribData();

	if ( attrib.getDataType() != DataType::ReqMsgEnum )
	{
		EmaString temp( "Attempt to getReqMsg() while actual dataType is " );
		temp += getDTypeAsString( attrib.getDataType() );
		throwIueException( temp );
	}

	return static_cast<const ReqMsg&>( attrib );
}

const RefreshMsg& Attrib::getRefreshMsg() const
{
	const Data& attrib = _pDecoder->getAttribData();

	if ( attrib.getDataType() != DataType::RefreshMsgEnum )
	{
		EmaString temp( "Attempt to getRefreshMsg() while actual dataType is " );
		temp += getDTypeAsString( attrib.getDataType() );
		throwIueException( temp );
	}

	return static_cast<const RefreshMsg&>( attrib );
}

const UpdateMsg& Attrib::getUpdateMsg() const
{
	const Data& attrib = _pDecoder->getAttribData();

	if ( attrib.getDataType() != DataType::UpdateMsgEnum )
	{
		EmaString temp( "Attempt to getUpdateMsg() while actual dataType is " );
		temp += getDTypeAsString( attrib.getDataType() );
		throwIueException( temp );
	}

	return static_cast<const UpdateMsg&>( attrib );
}

const StatusMsg& Attrib::getStatusMsg() const
{
	const Data& attrib = _pDecoder->getAttribData();

	if ( attrib.getDataType() != DataType::StatusMsgEnum )
	{
		EmaString temp( "Attempt to getStatusMsg() while actual dataType is " );
		temp += getDTypeAsString( attrib.getDataType() );
		throwIueException( temp );
	}

	return static_cast<const StatusMsg&>( attrib );
}

const PostMsg& Attrib::getPostMsg() const
{
	const Data& attrib = _pDecoder->getAttribData();

	if ( attrib.getDataType() != DataType::PostMsgEnum )
	{
		EmaString temp( "Attempt to getPostMsg() while actual dataType is " );
		temp += getDTypeAsString( attrib.getDataType() );
		throwIueException( temp );
	}

	return static_cast<const PostMsg&>( attrib );
}

const AckMsg& Attrib::getAckMsg() const
{
	const Data& attrib = _pDecoder->getAttribData();

	if ( attrib.getDataType() != DataType::AckMsgEnum )
	{
		EmaString temp( "Attempt to getAckMsg() while actual dataType is " );
		temp += getDTypeAsString( attrib.getDataType() );
		throwIueException( temp );
	}

	return static_cast<const AckMsg&>( attrib );
}

const GenericMsg& Attrib::getGenericMsg() const
{
	const Data& attrib = _pDecoder->getAttribData();

	if ( attrib.getDataType() != DataType::GenericMsgEnum )
	{
		EmaString temp( "Attempt to getGenericMsg() while actual dataType is " );
		temp += getDTypeAsString( attrib.getDataType() );
		throwIueException( temp );
	}

	return static_cast<const GenericMsg&>( attrib );
}

const FieldList& Attrib::getFieldList() const
{
	const Data& attrib = _pDecoder->getAttribData();

	if ( attrib.getDataType() != DataType::FieldListEnum )
	{
		EmaString temp( "Attempt to getFieldList() while actual dataType is " );
		temp += getDTypeAsString( attrib.getDataType() );
		throwIueException( temp );
	}

	return static_cast<const FieldList&>( attrib );
}

const ElementList& Attrib::getElementList() const
{
	const Data& attrib = _pDecoder->getAttribData();

	if ( attrib.getDataType() != DataType::ElementListEnum )
	{
		EmaString temp( "Attempt to getElementList() while actual dataType is " );
		temp += getDTypeAsString( attrib.getDataType() );
		throwIueException( temp );
	}

	return static_cast<const ElementList&>( attrib );
}

const Map& Attrib::getMap() const
{
	const Data& attrib = _pDecoder->getAttribData();

	if ( attrib.getDataType() != DataType::MapEnum )
	{
		EmaString temp( "Attempt to getMap() while actual dataType is " );
		temp += getDTypeAsString( attrib.getDataType() );
		throwIueException( temp );
	}

	return static_cast<const Map&>( attrib );
}

const Vector& Attrib::getVector() const
{
	const Data& attrib = _pDecoder->getAttribData();

	if ( attrib.getDataType() != DataType::VectorEnum )
	{
		EmaString temp( "Attempt to getVector() while actual dataType is " );
		temp += getDTypeAsString( attrib.getDataType() );
		throwIueException( temp );
	}

	return static_cast<const Vector&>( attrib );
}

const Series& Attrib::getSeries() const
{
	const Data& attrib = _pDecoder->getAttribData();

	if ( attrib.getDataType() != DataType::SeriesEnum )
	{
		EmaString temp( "Attempt to getSeries() while actual dataType is " );
		temp += getDTypeAsString( attrib.getDataType() );
		throwIueException( temp );
	}

	return static_cast<const Series&>( attrib );
}

const FilterList& Attrib::getFilterList() const
{
	const Data& attrib = _pDecoder->getAttribData();

	if ( attrib.getDataType() != DataType::FilterListEnum )
	{
		EmaString temp( "Attempt to getFilterList() while actual dataType is " );
		temp += getDTypeAsString( attrib.getDataType() );
		throwIueException( temp );
	}

	return static_cast<const FilterList&>( attrib );
}

const OmmOpaque& Attrib::getOpaque() const
{
	const Data& attrib = _pDecoder->getAttribData();

	if ( attrib.getDataType() != DataType::OpaqueEnum )
	{
		EmaString temp( "Attempt to getOpaque() while actual dataType is " );
		temp += getDTypeAsString( attrib.getDataType() );
		throwIueException( temp );
	}

	return static_cast<const OmmOpaque&>( attrib );
}

const OmmXml& Attrib::getXml() const
{
	const Data& attrib = _pDecoder->getAttribData();

	if ( attrib.getDataType() != DataType::XmlEnum )
	{
		EmaString temp( "Attempt to getXml() while actual dataType is " );
		temp += getDTypeAsString( attrib.getDataType() );
		throwIueException( temp );
	}

	return static_cast<const OmmXml&>( attrib );
}

const OmmAnsiPage& Attrib::getAnsiPage() const
{
	const Data& attrib = _pDecoder->getAttribData();

	if ( attrib.getDataType() != DataType::AnsiPageEnum )
	{
		EmaString temp( "Attempt to getAnsiPage() while actual dataType is " );
		temp += getDTypeAsString( attrib.getDataType() );
		throwIueException( temp );
	}

	return static_cast<const OmmAnsiPage&>( attrib );
}

File: Java/Ema/Core/src/main/java/com/rtsdk/ema/access/Attrib.java
///*|-----------------------------------------------------------------------------
// *|            This source code is provided under the Apache 2.0 license      --
// *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
// *|                See the project's LICENSE.md for details.                  --
// *|           Copyright (C) 2019 Refinitiv. All rights reserved.            --
///*|-----------------------------------------------------------------------------

package com.rtsdk.ema.access;


/**
 * Attrib is read only and is used for decoding only.
 * 
 * <p>The following code snippet shows extracting of Attrib and its content while processing RefreshMsg.</p>
 * <pre>
 * class AppClient implements OmmConsumerClient
 * {
 *     ...
 *
 *     public void onRefreshMsg( RefreshMsg refreshMsg, OmmConsumerEvent Event )
 *     {
 *         Attrib attrib = refreshMsg.attrib();
 *			
 *         switch( attrib.dataType() )
 *         {
 *         case DataTypes.FIELD_LIST :
 *            decode( attrib.fieldList() );
 *            break;
 *         case DataTypes.MAP :
 *            decode( attrib.map() );
 *            break;
 *         case DataTypes.NO_DATA :
 *            break;
 *         }
 *     }
 *     
 *     ...
 *
 * };
 * </pre>
 * 
 * This class is designed to efficiently perform extracting of Attrib and its content.<br>
 * Objects of this class are intended to be short lived or rather transitional.<br>
 * Objects of this class are not cache-able.
 *
 *
 * @see ComplexType
 * @see ReqMsg
 * @see RefreshMsg
 * @see UpdateMsg
 * @see StatusMsg
 * @see GenericMsg
 * @see PostMsg
 * @see AckMsg
 * @see FieldList
 * @see ElementList
 * @see Map
 * @see Vector
 * @see Series
 * @see FilterList
 * @see OmmOpaque
 * @see OmmXml
 * @see OmmAnsiPage
 * @see OmmError
 * 
 */
public interface Attrib
{
	/**
	 * Returns the DataType of the contained data.<br>
	 * Return of {@link com.rtsdk.ema.access.DataType.DataTypes#NO_DATA} signifies no data present in Attrib.<br>
	 * Return of {@link com.rtsdk.ema.access.DataType.DataTypes#ERROR} signifies error while extracting content of Attrib.
	 * 
	 * @return data type of the contained object
	 */
	public int dataType();

	/**
	 * Returns the complex type based on the DataType.
	 * 
	 * @return ComplexType class reference to contained object
	 */
	public ComplexType data();

	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.rtsdk.ema.access.ReqMsg}
	 * 
	 * @return {@link com.rtsdk.ema.access.ReqMsg} class reference to contained object
	 */
	public ReqMsg reqMsg();

	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.rtsdk.ema.access.RefreshMsg}
	 * 
	 * @return {@link com.rtsdk.ema.access.RefreshMsg} class reference to contained object
	 */
	public RefreshMsg refreshMsg();

	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.rtsdk.ema.access.UpdateMsg}
	 * 
	 * @return {@link com.rtsdk.ema.access.UpdateMsg} class reference to contained object
	 */
	public UpdateMsg updateMsg();

	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.rtsdk.ema.access.StatusMsg}
	 * 
	 * @return {@link com.rtsdk.ema.access.StatusMsg} class reference to contained object
	 */
	public StatusMsg statusMsg();

	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.rtsdk.ema.access.PostMsg}
	 * 
	 * @return {@link com.rtsdk.ema.access.PostMsg} class reference to contained object
	 */
	public PostMsg postMsg();

	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.rtsdk.ema.access.AckMsg}
	 * 
	 * @return {@link com.rtsdk.ema.access.AckMsg} class reference to contained object
	 */
	public AckMsg ackMsg();

	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.rtsdk.ema.access.GenericMsg}
	 * 
	 * @return {@link com.rtsdk.ema.access.GenericMsg} class reference to contained object
	 */
	public GenericMsg genericMsg();

	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.rtsdk.ema.access.FieldList}
	 * 
	 * @return {@link com.rtsdk.ema.access.FieldList} class reference to contained object
	 */
	public FieldList fieldList();

	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.rtsdk.ema.access.ElementList}
	 * 
	 * @return {@link com.rtsdk.ema.access.ElementList} class reference to contained object
	 */
	public ElementList elementList();

	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.rtsdk.ema.access.Map}
	 * 
	 * @return {@link com.rtsdk.ema.access.Map} class reference to contained object
	 */
	public Map map();

	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.rtsdk.ema.access.Vector}
	 * 
	 * @return {@link com.rtsdk.ema.access.Vector} class reference to contained object
	 */
	public Vector vector();

	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.rtsdk.ema.access.Series}
	 * 
	 * @return {@link com.rtsdk.ema.access.Series} class reference to contained object
	 */
	public Series series();

	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.rtsdk.ema.access.FilterList}
	 * 
	 * @return {@link com.rtsdk.ema.access.FilterList} class reference to contained object
	 */
	public FilterList filterList();

	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.rtsdk.ema.access.OmmOpaque}
	 * 
	 * @return {@link com.rtsdk.ema.access.OmmOpaque} class reference to contained object
	 */
	public OmmOpaque opaque();

	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.rtsdk.ema.access.OmmXml}
	 * 
	 * @return {@link com.rtsdk.ema.access.OmmXml} class reference to contained object
	 */
	public OmmXml xml();

	/**
	 * Returns the current OMM data represented as a specific complex type.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.rtsdk.ema.access.OmmAnsiPage}
	 * 
	 * @return {@link com.rtsdk.ema.access.OmmAnsiPage} class reference to contained object
	 */
	public OmmAnsiPage ansiPage();

	/**
	 * Returns Error.
	 * 
	 * @throws OmmInvalidUsageException if contained object is not {@link com.rtsdk.ema.access.OmmError}
	 * 
	 * @return {@link com.rtsdk.ema.access.OmmError} class reference to contained object
	 */
	public OmmError error();
}